experimental/authclient/oauth2.go

```diff
@@ -0,0 +1,97 @@
+package authclient
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend/httpclient"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	"golang.org/x/oauth2/jwt"
+)
+
+// OAuth2Type defines type of oauth2 grant type
+type OAuth2Type string
+
+const (
+	OAuth2TypeClientCredentials OAuth2Type = "client_credentials"
+	OAuth2TypeJWT               OAuth2Type = "jwt"
+)
+
+// OAuth2Options defines options for OAuth2 Client
+type OAuth2Options struct {
+	OAuth2Type     OAuth2Type
+	TokenURL       string
+	Scopes         []string
+	ClientID       string
+	ClientSecret   string
+	EndpointParams map[string]string
+	Subject        string
+	Email          string
+	PrivateKey     []byte
+	PrivateKeyID   string
+}
+
+func getOAuth2Client(client *http.Client, oAuth2Options OAuth2Options) (o *http.Client, err error) {
+	if client == nil {
+		client, err = httpclient.New(httpclient.Options{})
+		if err != nil {
+			return nil, err
+		}
+	}
+	switch oAuth2Options.OAuth2Type {
+	case OAuth2TypeClientCredentials:
+		return getOAuth2ClientCredentialsClient(*client, oAuth2Options)
+	case OAuth2TypeJWT:
+		return getOAuth2JWTClient(*client, oAuth2Options)
+	}
+	return client, fmt.Errorf("invalid/empty oauth2 type (%s)", oAuth2Options.OAuth2Type)
+}
+
+func getOAuth2ClientCredentialsClient(client http.Client, options OAuth2Options) (*http.Client, error) {
```

Comment:
nit: since you are using the pointer, the parameter here can be:

```suggestion
func getOAuth2ClientCredentialsClient(client *http.Client, options OAuth2Options) (*http.Client, error) {
```